In [1]:
# Get imports
import pandas as pd
import numpy as np 



In [2]:
# Let's make a function that takes in a database name as a string
# this function also performs our imports from env
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'

In [3]:
sql = """
    SELECT * from employees limit 3000
"""

# 5
Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

    - NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:amysql.pymysql

b. Intentionally make an error in your SQL query. What does the error message look like?

    - ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'fro employees limit 3000' at line 1")
[SQL: 
    SELECT * fro employees limit 3000
]
(Background on this error at: http://sqlalche.me/e/14/f405)

# 6 
Read the employees and titles tables into two separate DataFrames.

In [4]:
query_6_1 = '''
SELECT * 
FROM employees
'''

query_6_2 = '''
SELECT * 
FROM titles
'''

In [5]:
url = get_db_url("employees")
df_6_employees = pd.read_sql(query_6_1, url)
df_6_employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [6]:
df_6_titles = pd.read_sql(query_6_2, url)
df_6_titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

# 7 
How many rows and columns do you have in each DataFrame? Is that what you expected?

    - Employees 300,0024 rows x 6 columns
    - Titles 443,308 rows x 4 columns 
    Yes, I was expecting the titles table to have more records because one person can habe multiple jobs 
    



# 8 
Display the summary statistics for each DataFrame

In [7]:
df_6_employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [8]:
df_6_titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

# 9 
How many unique titles are in the titles DataFrame?

In [9]:
df_6_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [10]:
df_6_titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

# 10
What is the oldest date in the to_date column?

In [11]:
df_6_titles.to_date.min()

datetime.date(1985, 3, 1)

# 11
What is the most recent date in the to_date column?


In [12]:
df_6_titles.to_date.max()

datetime.date(9999, 1, 1)

# 

#  <center>Exercises II</center></h1>

# 2.1
Copy the users and roles DataFrames from the examples above.

In [13]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [14]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

# 2.2 
What is the result of using a right join on the DataFrames?

In [15]:
right_join = roles.merge(users, how='right')
right_join

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

# 2.3
What is the result of using an outer join on the DataFrames?

In [16]:
left_join =roles.merge(users, how='left')
left_join

id       name  role_id
0   1      admin      NaN
1   2     author      NaN
2   3   reviewer      NaN
3   4  commenter      NaN

# 2.4
What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [17]:
df.drop(columns=[“col1”, “col2”])

Empty DataFrame
Columns: [id, name, role_id]
Index: []

# 2. 5
Load the mpg dataset from PyDataset.

In [18]:
from pydataset import data
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

# 2.6
Output and read the documentation for the mpg dataset.

In [19]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




# 2.7
How many rows and columns are in the dataset?

There are 11 columns with 234 records

In [20]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


# 2.8
Check out your column names and perform any cleanup you may want on them  

In [25]:
mpg.columns

Index(['manufacturer', 'model', 'displ', 'year', 'cyl', 'trans', 'drv', 'cty',
       'hwy', 'fl', 'class'],
      dtype='object')

In [31]:
mpg.head(3)

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact

In [27]:
mpg.rename(columns={"cty": "city"})

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

# 2.9
Display the summary statistics for the dataset. 

In [33]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

# 2.10
How many different manufacturers are there? 

In [45]:
mpg['manufacturer'].unique()

array(['audi', 'chevrolet', 'dodge', 'ford', 'honda', 'hyundai', 'jeep',
       'land rover', 'lincoln', 'mercury', 'nissan', 'pontiac', 'subaru',
       'toyota', 'volkswagen'], dtype=object)

In [50]:
mpg['manufacturer'].value_counts()
brands = pd.DataFrame(mpg['manufacturer'].value_counts())
brands.shape

(15, 1)

# 2.11
How many different models are there?

In [54]:
mpg['model'].value_counts()
models = pd.DataFrame(mpg['model'].value_counts())
models.shape

(38, 1)

# 2.12
Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [58]:
mpg['mileage_difference'] = mpg[mpg.hwy - mpg.city]

AttributeError: 'DataFrame' object has no attribute 'city'

# 2.13
Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [ ]:
mpg['average_mileage'] =

# 2.14
Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission. 

In [ ]:
mpg['is_automatic'] =

#  2.15
Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average? 

# 2.16
Do automatic or manual cars have better miles per gallon? 

# 

 #  <center>Exercises III</center></h1>

# 3.1
Use your get_db_url function to help you explore the data from the chipotle database.

In [65]:
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'
    
url_c = get_db_url("chipotle")
sql_c = """
    SELECT * from orders

"""

url = get_db_url("employees")
df_c = pd.read_sql(sql_c, url_c)
df_c


id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

# 3.2
What is the total price for each order?

# 3.3
What are the most popular 3 items?

# 3.4
Which item has produced the most revenue?

# 3.5
Join the employees and titles DataFrames together.

# 3.6
For each title, find the hire date of the employee that was hired most recently with that title.

# 3.7
Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)   